{% extends "base.html" %}
{% from "macros.html" import showNavigationBar %}

{% block navigationBar %}
    {{ showNavigationBar('help') }}
{% endblock %}

{% block content %}
    <div class="indent">

    <div>
        <p>
            <i>Files that may be or may not be present are shown in italic.</i>
        </p>
        <h1>General</h1>
        <table>
            <tr>
                <th>Folder</th>
                <th>Filename</th>
                <th>Explanation</th>
            </tr>

            <tr>
                <td>Logs</td>
                <td>CAGECAT_version.txt</td>
                <td>The version of CAGECAT when your job was executed. Used for backtracking of errors.</td>
            </tr>

            <tr>
                <td>Logs</td>
                <td>{job_id}.log</td>
                <td>Log file which stores information about the executed analysis. Output from the executed programs/tools is stored in this file. May contain sanitization (from antiSmash tool) information. Can be inspected to get an impression of the steps that have been executed.</td>
            </tr>

            <tr>
                <td>Logs</td>
                <td>{job_id}_command.txt</td>
                <td>Command line interface command that was executed on CAGECAT's servers. Can be inspected to learn how submitted parameters are used in a command, executed using the command line interface.</td>
            </tr>

            <tr>
                <td>Logs</td>
                <td>{job_id}_options.txt</td>
                <td>Overview of submitted parameters and their values.</td>
            </tr>
        </table>
    </div>

<h1>cblaster</h1>

    <h2>Search</h2>

        <table>
            <tr>
                <th>Folder</th>
                <th>Filename</th>
                <th>Explanation</th>
            </tr>

            <tr>
                <td>Logs</td>
                <td><i>query_headers.csv</i></td>
                <td><i>Overview of headers of the submitted queries. Used when trying to recompute a job.</i></td>
            </tr>

            <tr>
                <td>Results</td>
                <td>{job_id}_binary.txt</td>
                <td>An absence/presence matrix of input queries vs. cluster hits. Shows the total number of hits in detected clusters per query sequence.</td>
            </tr>

            <tr>
                <td>Results</td>
                <td>{job_id}_plot.html</td>
                <td>Cluster heatmap visualization. Interactive HTML plot visualizing your input queries per scaffold hit.</td>
            </tr>

            <tr>
                <td>Results</td>
                <td>{job_id}_session.json</td>
                <td>Stores all information about found hit clusters and query sequences. Used when trying to execute a subsequent cblaster analysis.</td>
            </tr>

            <tr>
                <td>Results</td>
                <td>{job_id}_summary.txt</td>
                <td>Detailed overview of hit details/scores of input queries per scaffold hit.</td>
            </tr>

            {#            <tr>#}
            {#                <td>#}
            {#                    Intermediate genes?#}
            {#                </td>#}
            {#            </tr>#}

            <tr>
                <td>Uploads</td>
                <td><i>your_input_file.gbk</i></td>
                <td><i>The file you uploaded and what will be used as input file for the analysis.</i></td>
            </tr>

        </table>

    <h2>Recompute</h2>
        <p>The cblaster recompute module has the same output files as the cblaster search analysis.</p>

    <h2>Gene neighbourhood estimation</h2>
        <table>
            <tr>
                <th>Folder</th>
                <th>Filename</th>
                <th>Explanation</th>
            </tr>

            <tr>
                <td>Results</td>
                <td>{job_id}_plot.html</td>
                <td>Interactive HTML visualization of mean and median hit cluster sizes (bp) at different gap sizes and total clusters at different gap sizes. Helpful for a visual interpretation of robustness of cluster prediction.</td>
            </tr>

            <tr>
                <td>Results</td>
                <td>{job_id}_summary.txt</td>
                <td>Overview of GNE statistics</td>
            </tr>

        </table>

    <h2>Extract sequences</h2>
        <table>
            <tr>
                <th>Folder</th>
                <th>Filename</th>
                <th>Explanation</th>
            </tr>

            <tr>
                <td>Results</td>
                <td><i>{job_id}_sequences.txt</i></td>
                <td><i>Overview of where query genes hit on which scaffolds and positions. Is created when the "Download sequences" option was not checked.</i></td>
            </tr>

            <tr>
                <td>Results</td>
                <td><i>{job_id}_sequences.fasta</i></td>
                <td><i>Nucleotide FASTA of scaffold sequences where the query genes hit. Is created when the "Download sequences" option was checked.</i></td>
            </tr>

        </table>

    <h2>Extract clusters</h2>
        <table>
            <tr>
                <th>Folder</th>
                <th>Filename</th>
                <th>Explanation</th>
            </tr>

            <tr>
                <td>Results</td>
                <td>{prefix}{job_id}_cluster{number}.gbk</td>
                <td>GenBank files of selected cluster hits. {prefix} is only used when a value was entered for the "Prefix" option. {number} is inherent to the number of extracted clusters.</td>
            </tr>
        </table>

    <h2>Visualize cluster hits compared to queries</h2>
        <table>
            <tr>
                <th>Folder</th>
                <th>Filename</th>
                <th>Explanation</th>
            </tr>

            <tr>
                <td>Results</td>
                <td>{job_id}_plot.html</td>
                <td>Similar to a clinker plot, but comparisons are only available for your query cluster vs. individual hit cluster.</td>
            </tr>
        </table>

    <h1>clinker</h1>
        <table>
            <tr>
                <th>Folder</th>
                <th>Filename</th>
                <th>Explanation</th>
            </tr>


            <tr>
                <td>Results</td>
                <td>alignments.txt</td>
                <td>Overview of identity and similarity percentages between genes in clusters.</td>
            </tr>

            <tr>
                <td>Results</td>
                <td>{job_id}_plot.html</td>
                <td>The clinker visualisation enabling visual comparison of the submitted clusters and creation of such figures.</td>
            </tr>

            <tr>
                <td>Results</td>
                <td>{job_id}_session.json</td>
                <td>Stores all information about the cluster alignments.</td>
            </tr>
        </table>

    </div>
{% endblock %}
