{% extends "base.html" %}
{% from "macros.html" import showNavigationBar %}

{% block navigationBar %}
    {{ showNavigationBar('tutorial') }}
{% endblock %}

{% block content %}
    <div class="indent">
        <h1>Introduction</h1>
        <p>
            The CAGECAT web server facilitates sequence similarity searches and visualisations of microbial operons and
            gene clusters.It acts as a bridge between analyses executed by complex command line interfaces (CLIs) and
            aims to provide a user-friendly and intuitive interface to prevent difficult analyses execution for the
            individual
            researcher. Our service enables fast analyses using multiple interconnected tools, reducing the time
            required to execute such analyses. Researchers without programming experience encounter a steep learning
            curve to execute CLI analysis. Therefore, CAGECAT does not require any programming at all, lowering the
            technical barrier for researchers to execute gene cluster analyses. Additionally, connecting multiple gene
            cluster analysis tools on CAGECAT creates an online data analysis pipeline without the difficulties of
            connecting multiple individual tools, allowing users to combine the capabilities of each tool in
            customizable streamlined workflows.
        </p>

        <p>
            Executed jobs are stored on CAGECAT’s servers for 30 days and CAGECAT uses functional cookies only. This
            tutorial will guide you in submitting your first job at CAGECAT and will show you how you can search for
            homologous gene clusters for your protein of interest using a cblaster search analysis. In this tutorial,
            <b>bold sentences indicate steps for you to perform.</b>
        </p>

        <h2>Example workflow: starting an analysis</h2>
        <p>
            Two analyses (<a target="_blank" href="{{ url_for('tools.cblaster_search') }}">cblaster</a> and <a
                target="_blank" href="{{ url_for('tools.clinker') }}">clinker</a>) are implemented on CAGECAT. One can
            start an analysis by clicking the
            corresponding “Start” button on the home page (<a href="{{ url_for('tutorial') }}#fig1">Fig. 1</a>). Pages
            with additional explanation and publications
            of these tools are accessible by clicking the “Explanation” and “Publication” buttons, respectively. The
            valid input formats are shown per analysis. On CAGECAT, executed analyses are labelled as “jobs”. <b>Click
            the
            “Start” button located at the cblaster analysis.</b>
        </p>

        <div style="margin-left: 10%;">
            <img id="fig1" src="{{ url_for('static', filename='images/tutorial/tutorial_1_entry_points.PNG') }}" width="70%">
            <figcaption class="margin-top-1p">Figure 1</figcaption>
        </div>


        <h2>Selecting job parameters</h2>
        <p>
            You are now viewing the page for specification of parameters to be used in your analysis. The values for
            parameters you enter will be used when executing the analysis and will therefore have a direct impact on the
            results of your analysis. Input fields with a (*) are required. Input fields have restricted input (i.e.
            only numbers) and leaving a required input field blank will prevent you from submitting your job. At some
            analysis pages, job parameters (the values of settings to be used in your analysis) are split into basic and
            advanced parameters. More background information on the different types of analyses and explanations can be
            found <a target="_blank" href="{{ url_for('tools.tools_explanation') }}">here</a>.
        </p>

        <h3>Basic parameters</h3>
        <p>
            CAGECAT provides functionality to notify you by email when your job has finished executing. <b>You are able
            to provide a title for your job and enter your email address</b> for receiving a notification email. Your
            e-mail address is removed from our database after the notification e-mail has been sent. Providing an email
            address
            or job title is not required (<a href="{{ url_for('tutorial') }}#fig2">Fig. 2</a>). Note the <img
                class="help-button"
                src="{{ url_for('static', filename='images/question-circle.svg') }}"/>
            button shown for every parameter. <b>Click the <img class="help-button"
                                                                src="{{ url_for('static', filename='images/question-circle.svg') }}"/>
            button of the
            job title to show an explanation of the job title parameter.</b> The help texts are shown in the
            help-sidebar
            (right side of page) (<a href="{{ url_for('tutorial') }}#fig3">Fig. 3</a>). By default, the help-sidebar is
            shown, but can be hidden by clicking the <img
                src="{{ url_for('static', filename='images/tutorial/auxiliary/hide_help_sidebar.svg') }}"/>
            button. If the help side-bar is hidden, it can be shown by clicking the <img
                src="{{ url_for('static', filename='images/tutorial/auxiliary/show_help_sidebar.svg') }}"/> button.
        </p>

        <div style="margin-left: 10%;">
            <img id="fig2" src="{{ url_for('static', filename='images/tutorial/tutorial_2_job_description.PNG') }}"/>
            <figcaption class="margin-top-1p">Figure 2</figcaption>

            <img id="fig3" src="{{ url_for('static', filename='images/tutorial/tutorial_3_help_sidebar.PNG') }}"/>
            <figcaption class="margin-top-1p">Figure 3</figcaption>
        </div>

        <h4>Input section</h4>
        <p>
            Several input methods are available for a cblaster search: (I) Remote, which will compare your input
            proteins of interest against known protein sequences stored on NCBI servers; (II) HMM (Hidden Markov Model),
            which uses HMM profile identifiers (from <a href="https://pfam.xfam.org/">Pfam</a>) to search in
            genus-specific (Fungi and Bacteria are
            supported) databases on the CAGECAT servers; and (III) Remote + HMM, which will combine hits from NCBI
            servers and local HMM databases. Option III allows you to use both HMM profile identifiers and custom
            protein sequences as input for your cblaster analysis. Clicking on the desired search method (in the <a
                href="https://cagecat.bioinformatics.nl/tools/search#inputSection">Input
            section</a>) will show input fields corresponding to that search method.
        </p>

        <h4>Remote method</h4>
        <p>
            The Remote method can be used with two types of input: (I) a FASTA/GenBank file. The FASTA file can contain
            nucleotide or protein sequences and should contain at least two sequences. The GenBank (nucleotide
            sequences) should only contain nucleotide sequences of query genes. Full-genome GenBank files will cause an
            error. GenPept files are not supported. You can upload your input file by clicking the <img
                src="{{ url_for('static', filename='images/tutorial/auxiliary/query_file_button.PNG') }}"/>
            button. Your uploaded file will be rewritten to a new file in order to ensure a clean input for your analysis (the sanitization process). In order for this sanitization to be successful, be sure that your input file does not contain any special characters (if applicable, use an encoding of “UTF-8”). Currently, GenBank files created by SnapGene are known to raise errors during sanitization.
            The uploaded
            file will be used as input for your analysis and remain on CAGECAT’s servers for 30 days; (II) NCBI
            accession numbers representing a protein sequence. The protein sequence of the NCBI accession number will be
            used (i.e. accession numbers <a href="https://www.ncbi.nlm.nih.gov/protein/QBE85648.1/">QBE85648.1</a> and <a href="https://www.ncbi.nlm.nih.gov/protein/QBE85649.1/">QBE85649.1</a>). If you wish to use the associated nucleotide
            sequences of your proteins of interest, you are kindly requested to download the associated nucleotide
            sequence files (on web page of NCBI, add all accessions of interest to your clipboard, navigate to the
            clipboard, select option “Send to” → “File” → “FASTA CDS”) via NCBI and use this as input file for your
            analysis on CAGECAT. <b>Click the “NCBI entries” (<img
                src="{{ url_for('static', filename='images/tutorial/auxiliary/ncbi_entries_checkbox.PNG') }}"/>) option to change the input type</b>. All NCBI identifiers
            should be separated by a new line (one accession number per line). Remote searches are dependent on the load
            at NCBI services, and as such, could vary in completion time.
        </p>

        <h4>Search section (<a href="{{ url_for('tutorial') }}#fig4">Fig. 4</a>)</h4>
        <p>
            The Entrez query allows for a quick filtering of organisms (e.g. only include hits from Aspergillus genus,
            achieved by entering “Aspergillus[organism]” in this input field.) Note that any typographical errors can
            have a major impact on results. The database parameter shows the remote databases to search in. Normally,
            leaving the default setting (“nr”) will be sufficient. The “Maximum hits” parameter allows you to limit the
            number of hits your search will return. Lowering this value may result in missed hits.
        </p>

        <div style="margin-left: 10%;">
            <img id="fig4" src="{{ url_for('static', filename='images/tutorial/tutorial_4_search_section.PNG') }}"/>
            <figcaption class="margin-top-1p">Figure 4</figcaption>
        </div>

        <h4>HMM method</h4>
        <p>
            <b>Click the “HMM” option in the “Input” section.</b> Parameters specific for HMM-based searches appear, and
            parameters for Remote-based searches disappear (the “Search” section) In general, selecting a different
            input type or search method leads to specific parameters appearing and some other parameters disappearing.
            As this option does not use the parameters in the “Search” section, this section is hidden. HMM searches use
            HMM profile identifiers from Pfam (e.g. “PF01262”; <a href="https://pfam.xfam.org/family/PF01262">Pfam
            link</a>). The HMM corresponding to this identifier will
            be used to find hits within local genus-specific databases (available for Fungi and Bacteria). It is
            obligatory to select a genus to search in. Not all genera are available to reduce storage costs as we wish
            to keep CAGECAT freely accessible. If you wish to have an additional genus-specific database available for
            your searches, you are able to submit your wishes <a href="{{ url_for('help_page') }}">here</a>.
            If it is a highly requested genus or you are a
            high-frequent user, we might add this genus to our available databases.
        </p>

        <h3>Advanced parameters</h3>
        <p>
            The advanced section of parameters is hidden by default, but can be visualised by clicking on <img
                src="{{ url_for('static', filename='images/tutorial/auxiliary/show_advanced_section.svg') }}"/>. <b>Click
            on <img src="{{ url_for('static', filename='images/tutorial/auxiliary/show_advanced_section.svg') }}"/> to
            show advanced parameters.</b> You are able to fully execute an analysis without touching the advanced
            parameters. In that case, the default values for these parameters are used. We will not go over every
            parameter as their usage and function can be read by clicking the corresponding <img class="help-button"
                                                                                                 src="{{ url_for('static', filename='images/question-circle.svg') }}"/>
            boxes. Major options are described below:
        </p>

        <p>
            The “Required sequences” parameter enables selection of the input sequences which <u>must</u> be included in
            the
            output hits (i.e. the core gene of your cluster). The headers of your input sequences (from input file) or
            entered NCBI accessions will be shown here and can be selected by clicking on them. Upon selection, the
            background of the sequence header will become blue or grey. Multiple sequences can be selected by holding
            CTRL while clicking on the desired required sequences (<a href="{{ url_for('tutorial') }}#fig5">Fig. 5</a>).
        </p>

        <div style="margin-left: 10%;">
            <img id="fig5" src="{{ url_for('static', filename='images/tutorial/tutorial_5_clustering_section.PNG') }}"/>
            <figcaption class="margin-top-1p">Figure 5</figcaption>
        </div>

        <p>
            By default, in the “Binary table” section, the “Hit attribute” is disabled. By selecting a different value
            for the “Key function” parameter, this parameter becomes available and can be changed (<a
                href="{{ url_for('tutorial') }}#fig6">Fig. 6</a>). In general,
            parameters that are not changeable are disabled and will not be used in the analysis. As another example,
            checking the “Find intermediate genes” checkbox in the “Intermediate genes” section will enable the other
            parameters in this section to be used in the analysis.
        </p>

        <div style="margin-left: 10%;">
            <img id="fig6"
                 src="{{ url_for('static', filename='images/tutorial/tutorial_6_additional_sections.PNG') }}"/>
            <figcaption class="margin-top-1p">Figure 6</figcaption>
        </div>

        <h3>Example input</h3>
        <p>
            On the help-sidebar located on the right side of the page, the “Example input” button will enter predefined
            values in the input fields. <b>Click the “Example input” button (<a href="{{ url_for('tutorial') }}#fig3">Fig.
            3</a>).</b> Notice how the “NCBI entries”
            option
            is selected in the “Input” section and NCBI identifiers have been entered.
        </p>

        <h2>Submitting your job</h2>
        <p>
            If you have finished changing the values for the parameters, you are ready to submit your job by clicking
            the “Submit” button at the bottom of the analysis page (<a href="{{ url_for('tutorial') }}#fig6">Fig. 6</a>).
            <b>Enter your email address to get notified
                when your job has finished, and click the “Submit” button to submit your job.</b> Your job will be
            placed in
            the
            queue to be executed. If there is no queue, your job will start immediately. Your input protein sequences
            and entered parameter values will be sent to CAGECAT’s servers and your analysis will be executed.
            Meanwhile, the status of your job is shown and will be updated every few seconds (<a
                href="{{ url_for('tutorial') }}#fig7">Fig. 7</a>). This status page
            differs per job type, as every job has different steps that are executed.
        </p>

        <div style="margin-left: 10%;">
            <img id="fig7" src="{{ url_for('static', filename='images/tutorial/tutorial_7_job_status_update.PNG') }}"/>
            <figcaption class="margin-top-1p">Figure 7</figcaption>
        </div>

        <h2>Viewing the analysis results</h2>
        <p>
            <i>Note: to view an example output, the “Example output” button will redirect you to a page with example
                results of an analysis (<a href="{{ url_for('tutorial') }}#fig3">Fig. 3</a>)</i>
        </p>

        <p>
            The results page differs per executed analysis. <a href="{{ url_for('tutorial') }}#fig8">Fig. 8</a> shows
            the output page of a cblaster search analysis.
            However, the title (if given), available subsequent analyses and download/bookmark buttons are shown on
            every analysis results page.
        </p>

        <div style="margin-left: 10%;">
            <img id="fig8" src="{{ url_for('static', filename='images/tutorial/tutorial_8_results_page.PNG') }}" width="80%"/>
            <figcaption>
                Figure 8: overview of results page of a cblaster search analysis. 1: job title (if given); 2:
                buttons to download results to your computer and bookmark this page in your browser; 3: buttons to
                execute subsequent
                analyses. The available subsequent analyses per job type vary, and possible workflows can be inspected
                <a
                        href="https://cagecat.bioinformatics.nl/tools/#possibleWorkflows">here</a>.
                The (selected) output from the current job will be used as input for the subsequent job you select; 4:
                returned cluster hits from your search which are currently unselected to be used in subsequent analysis;
                5:
                selected cluster hits that will be used in subsequent analysis; 6: buttons to move cluster hits from the
                “unselected” pane to the “selected” pane. The selected cluster hits will be used in subsequent analysis;
                7:
                total number of selected clusters and queries. Input queries can be shown by clicking on the “Select
                queries” button (likewise, the cluster hits can be shown by clicking on the "Select clusters" button). The selected queries will be used in the “Extract sequences” analysis to filter for clusters that have these query genes.
            </figcaption>
        </div>

        <p>
            If an analysis produces a plot (HTML format), this is shown directly in your web browser. For a cblaster
            search analysis, the identified cluster hits are visualised (<a href="{{ url_for('tutorial') }}#fig9">Fig.
            9</a>). <b>Move the top 4 cluster hits to the
            “selected” pane (right pane).</b>
        </p>

        <div style="margin-left: 10%;">
            <img id="fig9"
                 src="{{ url_for('static', filename='images/tutorial/tutorial_9_results_page_html_output.PNG') }}"/>
            <figcaption class="margin-top-1p">Figure 9</figcaption>
        </div>

        <h2>Downloading results</h2>
        <p>
            <b>Click on the “Download results” button on the output page, shown in <a
                    href="{{ url_for('tutorial') }}#fig8">Fig. 8</a> (number 2).</b> Save the file
            to your
            computer and open it. Three folders are present (“logs”, “results”, “uploads”). In the latter, you will find
            any files you uploaded on the parameter page of your analysis. In the “logs”' folder, several files, named
            with your job ID, are present for inspection of how your job was executed and possible backtracking of
            errors. The command executed on the CAGECAT server is logged in the <span
                class="courier">job_id_command.txt</span> file, to give you an
            indication how your analysis was executed (<b>exercise: examine how the parameters and the values you
            entered
            on the parameters page are used in the command</b>). Your submitted parameter values are logged in the
            <span class="courier">job_id_options.txt</span> file and the log of execution of your job can be read in the
            file named <span class="courier">job_id.log</span>. In
            the “results” folder several other files are present of which a detailed explanation can be found <a
                href="{{ url_for('output_files_explanation') }}">here</a>.
        </p>

        <p>
            <i>Note: dependent on the analysis you executed, other files may be or may not be present.</i>
        </p>

        <h2>Executing a subsequent analysis</h2>
        <p>
            Depending on the analysis you executed, several subsequent analyses are available. An overview of possible
            workflows can be found <a href="https://cagecat.bioinformatics.nl/tools/#possibleWorkflows">here</a>. <b>Click
            the “Extract clusters” button and validate that the selected clusters are
            shown on the parameter page of the “Extract clusters” analysis (<a href="{{ url_for('tutorial') }}#fig10">Fig.
                10</a>).</b> If you did not select any of
            the
            cluster hits on the previous page, all clusters are selected by default (example shown in <a
                href="{{ url_for('tutorial') }}#fig11">Fig. 11</a>).
        </p>

        <div style="margin-left: 10%;">
            <img id="fig10"
                 src="{{ url_for('static', filename='images/tutorial/tutorial_10_extract_clusters_selected_clusters.PNG') }}"/>
            <figcaption class="margin-top-1p">Figure 10</figcaption>

            <img id="fig11"
                 src="{{ url_for('static', filename='images/tutorial/tutorial_11_all_clusters_selected.PNG') }}"/>
            <figcaption class="margin-top-1p">Figure 11</figcaption>
        </div>

        <p>
            After the “Extract clusters” job has finished executing, you are able to download your results. However,
            this analysis produced GenBank files that contain the protein sequences per cluster hits. Subsequently, you
            are able to visualise all cluster hits via clinker (the subsequent analysis), or download all GenBank files
            and use these in another analysis.
        </p>

        <h1>Auxiliary functionality</h1>
        <h2>Submitting feedback</h2>
        <p>
            You are able to submit feedback by navigating <a href="{{ url_for('help_page') }}">here</a>.
        </p>

        <h2>Viewing results of an executed job</h2>
        <p>
            You are able to look up the results of a previous job by clicking on the “Results” pane in the navigation
            bar at the top of every web page (<a href="{{ url_for('tutorial') }}#fig12">Fig. 12</a>), by browsing <a
                href="{{ url_for('result.result_from_job_id') }}">here</a> and entering your job ID, or clicking on the
            navigational links shown in the bar on the left side of every page (<a
                href="{{ url_for('tutorial') }}#fig13">Fig. 13</a>). These navigational links are
            stored locally after you submit a job.

        </p>

        <div style="margin-left: 10%;">
            <div>
                <img id="fig12"
                     src="{{ url_for('static', filename='images/tutorial/tutorial_12_navigation_bar.PNG') }}"/>
                <figcaption class="margin-top-1p">Figure 12</figcaption>
            </div>
            <br/>

            <img id="fig13" src="{{ url_for('static', filename='images/tutorial/tutorial_13_server_status.PNG') }}"/>
            <figcaption class="margin-top-1p">Figure 13</figcaption>
        </div>
    </div>
{% endblock %}
